### ISSUE TITLE
Come up with a release cycle

### ISSUE THREAD

**Issue by josegonzalez** [python-diamond/Diamond#181]
We should come up with a release cycle:

- Monthly? Bi-monthly?
- Should we concentrate on features/bugs for releases?
- How do we prioritize existing PRs/issues?
- Who has access to creating a pypi package?
- Should we also provide deb packages as part of a release?

This is meant as an RFC. Plenty of people use diamond, and it sucks that we can't give them a "known good" version at any point.

**Comment by mattrobenolt** [comment]
I have access to pypi. But we should definitely provide debs.

**Comment by mattrobenolt** [comment]
My biggest problem with all of this is that all the version bumping and changes literally only affect all the collectors.

I'd like to separate out the collectors from the core Diamond library. That way collectors can be managed independently instead of a core "THIS IS ALL OF THE THINGS".

I'd like to do a plugin system like we do for Sentry, then leave some of the collectors out of core and make it easy to install plugins and the community can maintain.  (tangential to this issue)

**Comment by mattrobenolt** [comment]
But I think doing that split will help with the prioritization issue as well. There are many issues that I leave untouched because I have 0 context on them, and most are just, "ehhhhh, this looks reasonable. Does it have tests? Cool. Merge."

Which is not the best for reviewing and is stressful.

**Comment by josegonzalez** [comment]
I feel the same way - and lots of the collectors seem to make semi-questionable changes, so its difficult to know whats what.

We should *also* have a standard release process for the collectors, and ensure we have an easy installation step for them. Maybe release the core collectors - cpu, disk, mem - with diamond and split everything else into separate projects?

**Comment by mattrobenolt** [comment]
Yeah, that was my thoughts exactly. It'd help development on core without dragging along this huge suite of collectors along with it. Then we'd be able to hand over some of the collectors to their own maintainers who actually have the right context.

Lemme hack on/draft up a proposal this weekend for what I think this should look like.

**Comment by mattrobenolt** [comment]
See: #183

**Comment by josegonzalez** [comment]
@kormoc I'm going to do releases twice a month from now on - assuming there are changes - on the first and the 15th. This should allow people to start saying "version x doesn't work". I'll make the first release tomorrow - @mattrobenolt I'm guessing you can push to pypi.

Once we get the ball rolling, we can start looking at bugs within diamond versus just collectors and run through them all :)

**Comment by eMerzh** [comment]
Hi guys, any news in the release process? 
i'm waiting for the next release of diamond :) 

Twice a month seems fair and really more user friendly than the unreleased-continuously-moving master

**Comment by Investocat** [comment]
:+1:

**Comment by mattrobenolt** [comment]
Just throwing in my two cents at this point:

I review and accept pull requests randomly, but I have very little time to actually contribute to Diamond. Let alone keep track of changes and publish proper releases.

I can schedule and run the script that cuts a new release and rely on the changelog being git commits, but that's about all I can offer.

Unfortunately, I'm not a fan of blindly doing that without testing, and that's not an effort that I can commit to at this point.

I get pinging privately about this issue, and it just makes me feel bad that I can't contribute more, but this project has kinda fallen into my lap since @kormoc has been unable to contribute.

I personally really like Diamond, but I'm actively moving Disqus away from it back to collectd just because I can't also be burdened with maintaining our metrics collection library. I have way too much on my plate to `actively` maintain and give this project the love that it needs and deserves.

So like I said, I can help out and do what I can, but doing proper releases are difficult. Need to comfortably maintain backwards compatibility, make sure that things are being versioned accordingly. So cutting new releases means reviewing all of the commits, etc and I'm not comfortable that I can make a good judgement call here.

To me, the biggest problem here is that Diamond keeps all of these crazy collectors in core, and I said this back in https://github.com/python-diamond/Diamond/issues/181#issuecomment-100332821. If Diamond were split out into core + plugins, it would make this process much much easier.

Then plugins could iterate at their own pace on their own terms without affecting the core Diamond service. Which would make code review and release process much simpler.

As it is today, it's an extremely difficult task reviewing code for collectors that I have 0 context on, for example, collectors for software or hardware that I don't use or interact with. On top of making sure whatever changes are made maintain backwards compatibility.

Again, if we can split out collectors into separate projects (we can use the `python-diamond` namespace), I'd be happy to get more committed to cutting actual releases, and this refactor can be planned for a `5.0` and get back on track.

Without doing this work, I'd be highly uncomfortable in my current position to contribute more at this point.

I hope this sheds some light into my perspective as a maintainer. :cry:

**Comment by justin8** [comment]
Would the idea with that be that the collector plugins are pip-intallable seperately and just depend on diamond itself?

**Comment by mattrobenolt** [comment]
Yeah, and we can manage plugins through the `entry_points` in plugins' `setup.py`.

**Comment by Investocat** [comment]
What's the blocker for the split? What needs to get done? Can we help out?

**Comment by mattrobenolt** [comment]
Nothing's really a blocker, just the work needs to be done. I'd be happy to help out.

The only thing that I don't have a clear grasp on is how this would translate to system packaging since Diamond is typically installed from a deb/rpm package.

**Comment by Investocat** [comment]
There's really just three options:

 * Per collector deb/rpms
 * Pull in collectors via git submodules
 * A mix of the two

I'm all for a rip the bandaid approach and just set it up as per collector packages.

How would we setup unit tests? Would the main diamond travis setup pull down all the collectors and run?

Will handlers also get moved out? They seem like another messy piece that shouldn't be in the core.

**Comment by mattrobenolt** [comment]
I'll have to think about it. I don't have all the answers at the moment off the top of my head.

The only thoughts I have are with regard to testing. I assume it'd be easy for us to provide a base "DiamondCollectorTest" class that all collectors would use, then they'd be tested in a similar fashion. A collector is just adhering to an API of the diamond server. So if we just provided a little class to assert that behavior, it'd be easy.

**Comment by justin8** [comment]
If you can give some direction, myself and I'm sure some others can help with the actual split. I don't usually do much deb packaging but could help with that. I'm already the maintainer on Arch, so splitting that one out will be easy. Making sure it works correctly via pip as well should be pretty high up on the list, otherwise any other distros like RHEL/Fedora won't have a fun time trying to use it.
